### DISCUSSION TITLE
issue: Schema validation using zod: Custom error message is not show instead generic error message is shown

### DISCUSSION THREAD

**Discussion by bicky-tmg** [react-hook-form/react-hook-form#9063] [Q&A] [answered]
### Version Number

^7.35.0

### Codesandbox/Expo snack

https://codesandbox.io/s/competent-ishizaka-mtfb6f?file=/src/Form.tsx

### Steps to reproduce

1. Go to 'https://codesandbox.io/s/competent-ishizaka-mtfb6f?file=/src/Form.tsx'
2. Click on 'Register account' button
3. Scroll down to 'console: to see the error object'
4. See error - From schema definition the custom error message i have defined is "You must accept terms and conditions", but instead "Invalid literal value, expected true" is always thrown as the error message for terms. 

Please manage to find time to look into the issue. 

### Expected behaviour

The custom error message i have defined in the schema definition "You must accept terms and conditions", should be thrown as message instead of the generic error message "Invalid literal value, expected true". 

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
please try to debug yourself for this sort of question: 

![Screen Shot 2022-09-19 at 7 41 03 pm](https://user-images.githubusercontent.com/10513364/190990747-c7a7ceec-11aa-4ac5-a45b-d9e6ba143736.png)

```tsx
resolver: async (data, context, options) => {
  // you can debug your validation schema here
  console.log('formData', data)
  console.log('validation result', await anyResolver(schema)(data, context, options))
  return anyResolver(schema)(data, context, options)
},
```

**Comment by leapful** [comment, marked as answer]
@bicky-tmg 

This is not an issue from react-hook-form. It's from library zod where "zod.literal" does not accept custom message via "invalid_type_error" option.

Switch to "errorMap" will solve the issue.

Working demo: https://codesandbox.io/s/falling-sky-iu5g6u?file=/src/Form.tsx

**Reply by bluebill1049** [reply]
thank you @leapful again 🙏

**Reply by bicky-tmg** [reply]
Thank you @leapful, yeah the issue was related to the zod library but still i had issues with handling such case so thank you. 🙏 🙇‍♂️

**Reply by freddiemixell** [reply]
You are my.literal(Hero)

**Reply by mohammadkhakshoor** [reply]
@leapful 
 what about file input (input type='file')   and   <select>otp 1 , opt 2,</select>
if possible , please he me these two

**Reply by andersoncampolina** [reply]
Great job mate. thanks to share it. But... and when I need to refine a item too?
    cpfCnpj: z.literal(true, {errorMap: () => ({ message: "Documento inválido" })}).refine(ValidaCpfCnpj),
not working here.

**Reply by andersoncampolina** [reply]
Got it solved using like this:
`    cpfCnpj: z.string({errorMap: () => ({ message: 'Cpf | Cnpj é obrigatório' })}).refine(ValidaCpfCnpj, { message: 'Documento inválido' }),
`
